### PR TITLE
[SPARK-24072][SQL] clearly define pushed filters

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

filters like parquet row group filter, which is actually pushed to the data source but still to be evaluated by Spark, should also count as `pushedFilters`.

## How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @rdblue @gengliangwang

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala [hunk 1]
```diff
@@ -237,11 +239,16 @@ object DataSourceV2Relation {
...
-
-        (nonConvertiblePredicates ++ unhandledPredicates.keys, pushedPredicates.keys.toSeq)
+        }.keys
+        // The filters which are marked as pushed to this data source
+        val pushedFilters = filterSupport.pushedFilters()
+        val pushedPredicates = translatedMap.filter { case (_, f) =>
+          pushedFilters.contains(f)
+        }.keys.toSeq
+
+        (nonConvertiblePredicates ++ unhandledPredicates, pushedPredicates)
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala]: maybe we can reverse the map and make it simpler:
```
        val translatedMap: Map[Filter, Expression] = filters.flatMap { p =>
          DataSourceStrategy.translateFilter(p).map(f => f -> p)
        }.toMap

        // Catalyst predicate expressions that cannot be converted to data source filters.
        val convertiblePredicates = translatedMap.values.toSet
        val nonConvertiblePredicates = filters.filterNot(convertiblePredicates.contains)

        // Data source filters that cannot be pushed down. An unhandled filter means
        // the data source cannot guarantee the rows returned can pass the filter.
        // As a result we must return it so Spark can plan an extra filter operator.
        val unhandledPredicates =
          filterSupport.pushFilters(translatedMap.keys.toArray).map(translatedMap)
        // The filters which are marked as pushed to this data source
        val pushedPredicates = filterSupport.pushedFilters().map(translatedMap)

        (nonConvertiblePredicates ++ unhandledPredicates, pushedPredicates)
```

Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala]: Or
```
        val translatedResults: Seq[(Option[Filter], Expression)] = filters.map { p =>
          (DataSourceStrategy.translateFilter(p), p)
        }

        // A map from original Catalyst expressions to corresponding translated data source
        // filters. If a predicate is not in this map, it means it cannot be pushed down.
        val translatedMap: Map[Filter, Expression] = translatedResults.collect {
          case (f, p) if f.isDefined => (f.get, p)
        }.toMap
        
        // Catalyst predicate expressions that cannot be converted to data source filters.
        val nonConvertiblePredicates = translatedResults.collect {
          case (f, p) if f.isEmpty => p
        }

        // Data source filters that cannot be pushed down. An unhandled filter means
        // the data source cannot guarantee the rows returned can pass the filter.
        // As a result we must return it so Spark can plan an extra filter operator.
        val unhandledPredicates =
          filterSupport.pushFilters(translatedMap.keys.toArray).map(translatedMap)
        // The filters which are marked as pushed to this data source
        val pushedPredicates = filterSupport.pushedFilters().map(translatedMap)

        (nonConvertiblePredicates ++ unhandledPredicates, pushedPredicates)
```

**Comment by gengliangwang** [comment]
retest this please.

**Comment by cloud-fan** [comment]
retest this please

**Comment by rdblue** [comment]
Thanks for working on this, @cloud-fan! I was thinking about needing it just recently so that data sources can delegate to Spark when needed.

I'll have a thorough look at it tomorrow, but one quick high-level question: should we make these residuals based on the input split instead?

Input splits might have different residual filters that need to be applied. For example, if you have a time range query, `ts > X`, and are storing data by day, then you know that when `day(ts) > day(X)` that `ts > X` *must* be true, but when `day(ts) = day(X)`, `ts > X` *might* be true. So for only some splits, when scanning the boundary day, you need to run the original filter, but not for any other splits.

Another use case for a per-split residual is when splits might be different file formats. Parquet allows pushing down filters, but Avro doesn't. In a mixed table format it would be great for Avro splits to return the entire expression as a residual, while Parquet splits do the filtering.

What do you think?

**Comment by cloud-fan** [comment]
This is a very good point! Unfortunately Spark SQL doesn't support change the physical plan at a per-split fashion, and I'd say this feature is non-trivial to implement, and needs a design doc.

After we have this feature, it might make sense to shift some of the mixin traits from `DataSourceReader` to `DataReaderFactory`, so that they can be applied per split.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala [hunk 2]
```diff
@@ -56,7 +56,7 @@ case class DataSourceV2Relation(
 
   lazy val (
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala]: // afterScanFilters: predicates that need to be evaluated after the scan.
// pushedFilters: predicates that will be pushed down and only evaluated in the underlying data sources.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala]: they can overlap(think about parquet row group filter), so I'd remove the `only` in the above comment.

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala]: uh... Then, we should mention it in the comment too. For example, 
```
// afterScanFilters: predicates that need to be evaluated after the scan.
// pushedFilters: predicates that will be pushed down and only evaluated in the underlying data sources.
// Note: afterScanFilters and pushedFilters can overlap. For example, the parquet row group filter
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala [hunk 3]
```diff
@@ -224,24 +226,29 @@ object DataSourceV2Relation {
         )
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala]: the indents between 221 and 223 look weird. Also correct them?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala [hunk 4]
```diff
@@ -224,24 +226,29 @@ object DataSourceV2Relation {
...
-        val translatedMap: Map[Expression, Filter] = filters.flatMap { p =>
-          DataSourceStrategy.translateFilter(p).map(f => p -> f)
-        }.toMap
+        val translateResults: Seq[(Option[Filter], Expression)] = filters.map { p =>
+          (DataSourceStrategy.translateFilter(p), p)
+        }
 
         // Catalyst predicate expressions that cannot be converted to data source filters.
-        val nonConvertiblePredicates = filters.filterNot(translatedMap.contains)
+        val nonConvertiblePredicates = translateResults.collect {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala]: Avoid using the `predicates` and `filters` as the variable names in the same function. This just makes the reader confused.

**Comment by gatorsmile** [comment]
LGTM except above comments. This PR does not change anything else but just clean the implementation of existing Data Source V2 API

**Comment by cloud-fan** [comment]
@rdblue After some more thoughts, I don't think Spark can support per-split physical plan changing in the near future. My best suggestion for mixed format data source is to have different implementations and union them at the end. We can add some kind of union API to data source v2 to make this thing easier.  What do you think?

**Comment by gatorsmile** [comment]
LGTM. Thanks! Merged to master.

**Comment by rdblue** [comment]
@cloud-fan, union doesn't really help. I already have support for mixed-formats working just fine. The format isn't the problem, it is filtering (and a similar problem with projection). Parquet allows you to push down filters, while Avro doesn't.

Right now, I'm running filters inside my data source to ensure that the result always matches pushed filters, which is okay but doesn't use codegen. Since we already have a need for per-split filters for residuals, we could do something similar in Spark instead of in the data sources and allow each split to return a residual. Then Spark would add a codegen'ed filter before proceeding with the rest of the physical plan. You might think of it as a `ResidualFilter` node, where the filter expression changes, instead of a separate physical plan.

**Comment by cloud-fan** [comment]
Hi @rdblue , you can also create a codegen-ed filter in your data source, but the real problem is whole-stage-codegen. The code is generated at the driver side, sent to and compiled at the executor side. It is non-trivial to change it to generating code at executor side. If union doesn't work, let's think more about this and write a design doc first. I'm willing to join the discussion and help to make a design, thanks!

**Comment by cloud-fan** [comment]
If we don't care about whole stage codegen, I think it's possible to dynamically change the filter condition at executors and codegen it.

Now the problem becomes, how would the data source API look like? I'd image some systems have to do filter pushdown at the driver side, so that they can connect to the metadata service only once. One possible way is to allow users to mixin `SupportsPushDownFilter` to both `DataSourceReader` and `DataReaderFactory`.

**Comment by rdblue** [comment]
@cloud-fan, here's a commit that demonstrates the idea and implementation: https://github.com/rdblue/spark/commit/b41eb1ef2af38c510f5426a096c586a93e4a5556

That adds `residualFilters` to the `InputPartition` and updates `DataSourceRDD` to compile and run the filter expressions if there are any.

**Pull request closed without merging**